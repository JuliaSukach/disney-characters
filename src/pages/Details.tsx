import { useState, useEffect } from 'react'
import axios from 'axios'
import { useParams } from 'react-router-dom'
import '../css/details.css'
import ErrorMessage from '../components/ErrorMessage'
import Spinner from '../components/Spinner'

const Details = () => {
    // Define the Character interface
    interface Character {
        _id: number
        name: string
        imageUrl: string
        films?: string[]
        tvShows?: string[]
        videoGames?: string[]
        parkAttractions?: string[]
        sourceUrl?: string
    }
    const { id } = useParams<{ id: string }>() // Get character ID from URL
    const [character, setCharacter] = useState<Character | null>(null)
    const [loading, setLoading] = useState(false)
    const [error, setError] = useState('')

    // Function to validate response structure
    const isValidCharacterResponse = (response: any) => {
        return response?.data?.data && response?.data?.info?.count > 0
    }

    useEffect(() => {
        // Fetch character details using the ID from the URL
        const fetchCharacter = async () => {
            setLoading(true)
            setError('')
            try {
                const response = await axios.get(`https://api.disneyapi.dev/character/${id}`)
                if (isValidCharacterResponse(response)) {
                    setCharacter(response.data.data)
                } else {
                    setError('Character not found.')
                }
            } catch (error) {
                setError('Error fetching character details. Please try again later.')
            } finally {
                setLoading(false)
            }
        }

        fetchCharacter()
    }, [id])

    if (loading) return <Spinner />
    if (error) return <ErrorMessage message={error} />

    return (
        <div className='details-container'>
            {character ? (
                <>
                    <div className='details-image-container'>
                        <img className="details-image" src={character.imageUrl  || '/images/character.png'} alt={character.name} />
                    </div>
                    <div className='details-content'>
                        <h1 className='details-name'>{character.name}</h1>

                        {/* Films Section */}
                        <h3>Films</h3>
                        {character.films && character.films.length > 0 ? (
                            <ul>
                                {character.films.map((film, index) => (
                                    <li key={index}>{film}</li>
                                ))}
                            </ul>
                        ) : (
                            <p>No films available.</p>
                        )}

                        {/* TV Shows Section */}
                        <h3>TV Shows</h3>
                        {character.tvShows && character.tvShows.length > 0 ? (
                            <ul>
                                {character.tvShows.map((show, index) => (
                                    <li key={index}>{show}</li>
                                ))}
                            </ul>
                        ) : (
                            <p>No TV shows available.</p>
                        )}

                        {/* Video Games Section */}
                        <h3>Video Games</h3>
                        {character.videoGames && character.videoGames.length > 0 ? (
                            <ul>
                                {character.videoGames.map((game, index) => (
                                    <li key={index}>{game}</li>
                                ))}
                            </ul>
                        ) : (
                            <p>No video games available.</p>
                        )}

                        {/* Park Attractions Section */}
                        <h3>Park Attractions</h3>
                        {character.parkAttractions && character.parkAttractions.length > 0 ? (
                            <ul>
                                {character.parkAttractions.map((attraction, index) => (
                                    <li key={index}>{attraction}</li>
                                ))}
                            </ul>
                        ) : (
                            <p>No park attractions available.</p>
                        )}

                        {/* External Source Link */}
                        {character.sourceUrl ? (
                            <div className='learn-more-section'>
                                <h3>Learn More</h3>
                                <a href={character.sourceUrl} target='_blank' rel='noopener noreferrer'>
                                    Visit Disney Fandom for more details
                                </a>
                            </div>
                        ) : (
                            <p>No external resources available.</p>
                        )}
                    </div>
                </>
            ) : (
                <p className='error-message'>No character details available.</p>
            )}
        </div>
    )
}

export default Details