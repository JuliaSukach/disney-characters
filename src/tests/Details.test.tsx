import { render, screen, waitFor } from '@testing-library/react'
import { MemoryRouter, Route, Routes } from 'react-router-dom'
import axios from 'axios'
import Details from '../pages/Details'

// Mock axios to avoid making real API requests
jest.mock('axios')
const mockedAxios = axios as jest.Mocked<typeof axios>

describe('Details Component', () => {
    const characterId = '123'
    const characterData = {
        _id: 123,
        name: 'Mickey Mouse',
        imageUrl: 'https://example.com/mickey.png',
        films: ['Steamboat Willie'],
        tvShows: ['The Mickey Mouse Clubhouse'],
        videoGames: ['Kingdom Hearts'],
        sourceUrl: 'https://disney.fandom.com/wiki/Mickey_Mouse',
    }

    beforeEach(() => {
        mockedAxios.get.mockClear()
    })

    test('shows loading spinner when fetching data', async () => {
        mockedAxios.get.mockResolvedValueOnce(new Promise(() => {})) // Simulate loading state

        const { asFragment } = render(
            <MemoryRouter initialEntries={[`/details/${characterId}`]}>
                <Routes>
                    <Route path='/details/:id' element={<Details />} />
                </Routes>
            </MemoryRouter>
        )

        // Assert loading spinner is displayed
        await waitFor(() => {
            expect(screen.getByRole('status', { name: /loading/i })).toBeInTheDocument()
        })
        // Take snapshot of the loading state
        expect(asFragment()).toMatchSnapshot()
    })

    test('renders character details successfully', async () => {
        mockedAxios.get.mockResolvedValueOnce({
            data: { data: characterData, info: { count: 1 } },
        })

        const { asFragment } = render(
            <MemoryRouter initialEntries={[`/details/${characterId}`]}>
                <Routes>
                    <Route path='/details/:id' element={<Details />} />
                </Routes>
            </MemoryRouter>
        )

        // Wait for the character details to load
        await waitFor(() => {
            expect(screen.getByText(characterData.name)).toBeInTheDocument()
        })

        // Assert all key data fields are present
        expect(screen.getByAltText(characterData.name)).toBeInTheDocument()
        expect(screen.getByText('Films')).toBeInTheDocument()
        expect(screen.getByText('Steamboat Willie')).toBeInTheDocument()
        expect(screen.getByText('TV Shows')).toBeInTheDocument()
        expect(screen.getByText('The Mickey Mouse Clubhouse')).toBeInTheDocument()
        expect(screen.getByText('Video Games')).toBeInTheDocument()
        expect(screen.getByText('Kingdom Hearts')).toBeInTheDocument()
        expect(screen.getByRole('link', { name: /Visit Disney Fandom for more details/i })).toBeInTheDocument()

        // Take snapshot of the rendered character details page
        expect(asFragment()).toMatchSnapshot()
    })

    test('displays error message if character not found', async () => {
        mockedAxios.get.mockResolvedValueOnce({
            data: { data: null, info: { count: 0 } },
        })

        const { asFragment } = render(
            <MemoryRouter initialEntries={[`/details/${characterId}`]}>
                <Routes>
                    <Route path='/details/:id' element={<Details />} />
                </Routes>
            </MemoryRouter>
        )

        // Wait for the "Character not found" message
        await waitFor(() => {
            expect(screen.getByText(/Character not found./i)).toBeInTheDocument()
        })

        // Snapshot of the not found state
        expect(asFragment()).toMatchSnapshot()
    })

    test('displays error message on fetch failure', async () => {
        mockedAxios.get.mockRejectedValueOnce(new Error('Network error'))

        const { asFragment } = render(
            <MemoryRouter initialEntries={[`/details/${characterId}`]}>
                <Routes>
                    <Route path='/details/:id' element={<Details />} />
                </Routes>
            </MemoryRouter>
        )

        // Wait for the error message to be displayed
        await waitFor(() => {
            expect(screen.getByText(/Error fetching character details. Please try again later./i)).toBeInTheDocument()
        })

        // Snapshot of the error state
        expect(asFragment()).toMatchSnapshot()
    })

    test('displays default image when character image is not available', async () => {
        const characterWithoutImage = { ...characterData, imageUrl: undefined }
        mockedAxios.get.mockResolvedValueOnce({
            data: { data: characterWithoutImage, info: { count: 1 } },
        })

        render(
            <MemoryRouter initialEntries={[`/details/${characterId}`]}>
                <Routes>
                    <Route path='/details/:id' element={<Details />} />
                </Routes>
            </MemoryRouter>
        )

        await waitFor(() => {
            expect(screen.getByAltText(characterWithoutImage.name)).toBeInTheDocument()
        })

        // Verify the fallback image is used when imageUrl is not available
        const imgElement = screen.getByAltText(characterWithoutImage.name) as HTMLImageElement
        expect(imgElement.src).toContain('/images/character.png')

        // Snapshot of fallback image scenario
        expect(imgElement).toMatchSnapshot()
    })
})
