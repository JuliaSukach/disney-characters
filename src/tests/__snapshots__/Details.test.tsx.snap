// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Details Component displays default image when character image is not available 1`] = `
<img
  alt="Mickey Mouse"
  class="details-image"
  src="/images/character.png"
/>
`;

exports[`Details Component displays error message if character not found 1`] = `
<DocumentFragment>
  <p
    class="error-message"
    role="alert"
    style="color: red;"
  >
    Character not found.
  </p>
</DocumentFragment>
`;

exports[`Details Component displays error message on fetch failure 1`] = `
<DocumentFragment>
  <p
    class="error-message"
    role="alert"
    style="color: red;"
  >
    Error fetching character details. Please try again later.
  </p>
</DocumentFragment>
`;

exports[`Details Component renders character details successfully 1`] = `
<DocumentFragment>
  <div
    class="details-container"
  >
    <div
      class="details-image-container"
    >
      <img
        alt="Mickey Mouse"
        class="details-image"
        src="https://example.com/mickey.png"
      />
    </div>
    <div
      class="details-content"
    >
      <h1
        class="details-name"
      >
        Mickey Mouse
      </h1>
      <h3>
        Films
      </h3>
      <ul>
        <li>
          Steamboat Willie
        </li>
      </ul>
      <h3>
        TV Shows
      </h3>
      <ul>
        <li>
          The Mickey Mouse Clubhouse
        </li>
      </ul>
      <h3>
        Video Games
      </h3>
      <ul>
        <li>
          Kingdom Hearts
        </li>
      </ul>
      <h3>
        Park Attractions
      </h3>
      <p>
        No park attractions available.
      </p>
      <div
        class="learn-more-section"
      >
        <h3>
          Learn More
        </h3>
        <a
          href="https://disney.fandom.com/wiki/Mickey_Mouse"
          rel="noopener noreferrer"
          target="_blank"
        >
          Visit Disney Fandom for more details
        </a>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Details Component shows loading spinner when fetching data 1`] = `
<DocumentFragment>
  <div
    aria-label="Loading"
    class="spinner"
    role="status"
  />
</DocumentFragment>
`;
